File: go/pgx/dsql/config_test.go
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package dsql

import (
	"os"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestConfigDefaults(t *testing.T) {
	cfg := Config{
		Host: "mycluster.dsql.us-east-1.on.aws",
	}

	resolved, err := cfg.resolve()
	require.NoError(t, err)

	assert.Equal(t, "mycluster.dsql.us-east-1.on.aws", resolved.Host)
	assert.Equal(t, "us-east-1", resolved.Region)
	assert.Equal(t, "admin", resolved.User)
	assert.Equal(t, "postgres", resolved.Database)
	assert.Equal(t, 5432, resolved.Port)
	assert.Equal(t, DefaultTokenDuration, resolved.TokenDuration)
}

func TestConfigDefaultTokenDuration(t *testing.T) {
	cfg := Config{
		Host: "mycluster.dsql.us-east-1.on.aws",
	}

	resolved, err := cfg.resolve()
	require.NoError(t, err)

	// Should default to 15 minutes
	assert.Equal(t, 15*time.Minute, resolved.TokenDuration)
}

func TestConfigExplicitRegion(t *testing.T) {
	cfg := Config{
		Host:   "mycluster.dsql.us-east-1.on.aws",
		Region: "eu-west-1", // Override parsed region
	}

	resolved, err := cfg.resolve()
	require.NoError(t, err)

	assert.Equal(t, "eu-west-1", resolved.Region)
}

func TestConfigClusterID(t *testing.T) {
	clusterID := "ijsamhssbh36dopuigphknejb4"

	cfg := Config{
		Host:   clusterID,
		Region: "us-west-2",
	}

	resolved, err := cfg.resolve()
	require.NoError(t, err)

	assert.Equal(t, clusterID+".dsql.us-west-2.on.aws", resolved.Host)
	assert.Equal(t, "us-west-2", resolved.Region)
}

func TestConfigClusterIDWithoutRegion(t *testing.T) {
	// Clear any environment region
	oldRegion := os.Getenv("AWS_REGION")
	oldDefaultRegion := os.Getenv("AWS_DEFAULT_REGION")
	os.Unsetenv("AWS_REGION")
	os.Unsetenv("AWS_DEFAULT_REGION")
	defer func() {
		if oldRegion != "" {
			os.Setenv("AWS_REGION", oldRegion)
		}
		if oldDefaultRegion != "" {
			os.Setenv("AWS_DEFAULT_REGION", oldDefaultRegion)
		}
	}()

	clusterID := "jbtgm4i7xmqphuo2mgamk7oeza"

	cfg := Config{
		Host: clusterID,
	}

	_, err := cfg.resolve()
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "region")
}

func TestConfigRegionFromEnv(t *testing.T) {
	oldRegion := os.Getenv("AWS_REGION")
	os.Setenv("AWS_REGION", "ap-northeast-1")
	defer func() {
		if oldRegion != "" {
			os.Setenv("AWS_REGION", oldRegion)
		} else {
			os.Unsetenv("AWS_REGION")
		}
	}()

	clusterID := "jyabtzxzk6utb27wod2toxeifm"

	cfg := Config{
		Host: clusterID,
	}

	resolved, err := cfg.resolve()
	require.NoError(t, err)

	assert.Equal(t, "ap-northeast-1", resolved.Region)
}

func TestConfigMissingHost(t *testing.T) {
	cfg := Config{}

	_, err := cfg.resolve()
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "host")
}

func TestConfigResolve(t *testing.T) {
	tests := []struct {
		name    string
		config  Config
		setup   func()    // Optional setup (e.g., set env vars)
		cleanup func()    // Optional cleanup
		wantErr bool
		errMsg  string
	}{
		{
			name:    "valid config with full hostname",
			config:  Config{Host: "mycluster.dsql.us-east-1.on.aws"},
			wantErr: false,
		},
		{
			name:    "valid config with host and port",
			config:  Config{Host: "mycluster.dsql.us-east-1.on.aws", Port: 5433},
			wantErr: false,
		},
		{
			name:    "valid config with cluster ID and region",
			config:  Config{Host: "ijsamhssbh36dopuigphknejb4", Region: "us-east-1"},
			wantErr: false,
		},
		{
			name:    "missing host",
			config:  Config{},
			wantErr: true,
			errMsg:  "host is required",
		},
		{
			name:    "empty host",
			config:  Config{Host: ""},
			wantErr: true,
			errMsg:  "host is required",
		},
		{
			name:    "port zero uses default",
			config:  Config{Host: "mycluster.dsql.us-east-1.on.aws", Port: 0},
			wantErr: false,
		},
		{
			name:    "port negative",
			config:  Config{Host: "mycluster.dsql.us-east-1.on.aws", Port: -1},
			wantErr: true,
			errMsg:  "port must be between 1 and 65535",
		},
		{
			name:    "port too high",
			config:  Config{Host: "mycluster.dsql.us-east-1.on.aws", Port: 70000},
			wantErr: true,
			errMsg:  "port must be between 1 and 65535",
		},
		{
			name:    "port at lower bound",
			config:  Config{Host: "mycluster.dsql.us-east-1.on.aws", Port: 1},
			wantErr: false,
		},
		{
			name:    "port at upper bound",
			config:  Config{Host: "mycluster.dsql.us-east-1.on.aws", Port: 65535},
			wantErr: false,
		},
		{
			name:   "cluster ID without region fails",
			config: Config{Host: "ijsamhssbh36dopuigphknejb4"},
			setup: func() {
				os.Unsetenv("AWS_REGION")
				os.Unsetenv("AWS_DEFAULT_REGION")
			},
			cleanup: func() {},
			wantErr: true,
			errMsg:  "region is required when host is a cluster ID",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Save and restore env vars if needed
			oldRegion := os.Getenv("AWS_REGION")
			oldDefaultRegion := os.Getenv("AWS_DEFAULT_REGION")
			defer func() {
				if oldRegion != "" {
					os.Setenv("AWS_REGION", oldRegion)
				} else {
					os.Unsetenv("AWS_REGION")
				}
				if oldDefaultRegion != "" {
					os.Setenv("AWS_DEFAULT_REGION", oldDefaultRegion)
				} else {
					os.Unsetenv("AWS_DEFAULT_REGION")
				}
			}()

			if tt.setup != nil {
				tt.setup()
			}
			if tt.cleanup != nil {
				defer tt.cleanup()
			}

			_, err := tt.config.resolve()
			if tt.wantErr {
				assert.Error(t, err)
				assert.Contains(t, err.Error(), tt.errMsg)
			} else {
				assert.NoError(t, err)
			}
		})
	}
}

func TestConfigInvalidPort(t *testing.T) {
	cfg := Config{
		Host: "mycluster.dsql.us-east-1.on.aws",
		Port: 70000,
	}

	_, err := cfg.resolve()
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "port")
}

func TestConfigAllFields(t *testing.T) {
	cfg := Config{
		Host:              "mycluster.dsql.us-east-1.on.aws",
		Region:            "us-east-1",
		User:              "myuser",
		Database:          "mydb",
		Port:              5433,
		Profile:           "myprofile",
		TokenDurationSecs: 120,
	}

	resolved, err := cfg.resolve()
	require.NoError(t, err)

	assert.Equal(t, "mycluster.dsql.us-east-1.on.aws", resolved.Host)
	assert.Equal(t, "us-east-1", resolved.Region)
	assert.Equal(t, "myuser", resolved.User)
	assert.Equal(t, "mydb", resolved.Database)
	assert.Equal(t, 5433, resolved.Port)
	assert.Equal(t, "myprofile", resolved.Profile)
	assert.Equal(t, 120*time.Second, resolved.TokenDuration)
}

func TestParseConnectionString(t *testing.T) {
	tests := []struct {
		name     string
		connStr  string
		expected Config
	}{
		{
			name:    "basic connection string",
			connStr: "postgres://admin@mycluster.dsql.us-east-1.on.aws/postgres",
			expected: Config{
				Host:     "mycluster.dsql.us-east-1.on.aws",
				User:     "admin",
				Database: "postgres",
			},
		},
		{
			name:    "with port",
			connStr: "postgres://admin@mycluster.dsql.us-east-1.on.aws:5433/mydb",
			expected: Config{
				Host:     "mycluster.dsql.us-east-1.on.aws",
				User:     "admin",
				Database: "mydb",
				Port:     5433,
			},
		},
		{
			name:    "with region parameter",
			connStr: "postgres://admin@mycluster.dsql.us-east-1.on.aws/postgres?region=eu-west-1",
			expected: Config{
				Host:     "mycluster.dsql.us-east-1.on.aws",
				User:     "admin",
				Database: "postgres",
				Region:   "eu-west-1",
			},
		},
		{
			name:    "with token duration",
			connStr: "postgres://admin@mycluster.dsql.us-east-1.on.aws/postgres?tokenDurationSecs=300",
			expected: Config{
				Host:              "mycluster.dsql.us-east-1.on.aws",
				User:              "admin",
				Database:          "postgres",
				TokenDurationSecs: 300,
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cfg, err := ParseConnectionString(tt.connStr)
			require.NoError(t, err)

			assert.Equal(t, tt.expected.Host, cfg.Host)
			assert.Equal(t, tt.expected.User, cfg.User)
			assert.Equal(t, tt.expected.Database, cfg.Database)
			if tt.expected.Port != 0 {
				assert.Equal(t, tt.expected.Port, cfg.Port)
			}
			if tt.expected.Region != "" {
				assert.Equal(t, tt.expected.Region, cfg.Region)
			}
			if tt.expected.TokenDurationSecs != 0 {
				assert.Equal(t, tt.expected.TokenDurationSecs, cfg.TokenDurationSecs)
			}
		})
	}
}

File: ruby/pg/test/unit/config_test.rb
# frozen_string_literal: true

# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

require "rspec"
require_relative "../../lib/aurora_dsql_pg"
require_relative "../../lib/aurora_dsql/pg/util"
require_relative "../../lib/aurora_dsql/pg/config"

RSpec.describe AuroraDsql::Pg::Config do
  describe "DEFAULTS" do
    it "has correct default values" do
      expect(AuroraDsql::Pg::Config::DEFAULTS[:user]).to eq("admin")
      expect(AuroraDsql::Pg::Config::DEFAULTS[:database]).to eq("postgres")
      expect(AuroraDsql::Pg::Config::DEFAULTS[:port]).to eq(5432)
      expect(AuroraDsql::Pg::Config::DEFAULTS[:max_lifetime]).to eq(55 * 60)
      expect(AuroraDsql::Pg::Config::DEFAULTS[:token_duration]).to eq(15 * 60)
      expect(AuroraDsql::Pg::Config::DEFAULTS[:pool_size]).to eq(5)
    end
  end

  describe "#resolve" do
    it "applies defaults for minimal config" do
      config = described_class.new(host: "mycluster.dsql.us-east-1.on.aws")
      resolved = config.resolve

      expect(resolved.host).to eq("mycluster.dsql.us-east-1.on.aws")
      expect(resolved.region).to eq("us-east-1")
      expect(resolved.user).to eq("admin")
      expect(resolved.database).to eq("postgres")
      expect(resolved.port).to eq(5432)
      expect(resolved.token_duration).to eq(15 * 60)
    end

    it "expands cluster ID to full hostname" do
      config = described_class.new(
        host: "ijsamhssbh36dopuigphknejb4",
        region: "us-west-2"
      )
      resolved = config.resolve

      expect(resolved.host).to eq("ijsamhssbh36dopuigphknejb4.dsql.us-west-2.on.aws")
      expect(resolved.region).to eq("us-west-2")
    end

    it "raises error for cluster ID without region" do
      ENV.delete("AWS_REGION")
      ENV.delete("AWS_DEFAULT_REGION")

      config = described_class.new(host: "ijsamhssbh36dopuigphknejb4")

      expect { config.resolve }.to raise_error(AuroraDsql::Pg::Error, /region is required/)
    end

    it "uses region from env for cluster ID" do
      ENV["AWS_REGION"] = "ap-northeast-1"
      config = described_class.new(host: "ijsamhssbh36dopuigphknejb4")
      resolved = config.resolve

      expect(resolved.region).to eq("ap-northeast-1")
      expect(resolved.host).to eq("ijsamhssbh36dopuigphknejb4.dsql.ap-northeast-1.on.aws")
    ensure
      ENV.delete("AWS_REGION")
    end

    it "raises error for missing host" do
      config = described_class.new

      expect { config.resolve }.to raise_error(AuroraDsql::Pg::Error, /host is required/)
    end

    it "raises error for invalid port" do
      config = described_class.new(
        host: "mycluster.dsql.us-east-1.on.aws",
        port: 70000
      )

      expect { config.resolve }.to raise_error(AuroraDsql::Pg::Error, /port must be between/)
    end

    it "allows explicit region override" do
      config = described_class.new(
        host: "mycluster.dsql.us-east-1.on.aws",
        region: "eu-west-1"
      )
      resolved = config.resolve

      expect(resolved.region).to eq("eu-west-1")
    end

    it "preserves all user-provided values" do
      config = described_class.new(
        host: "mycluster.dsql.us-east-1.on.aws",
        user: "myuser",
        database: "mydb",
        port: 5433,
        profile: "myprofile",
        token_duration: 300,
        pool_size: 10,
        application_name: "rails"
      )
      resolved = config.resolve

      expect(resolved.user).to eq("myuser")
      expect(resolved.database).to eq("mydb")
      expect(resolved.port).to eq(5433)
      expect(resolved.profile).to eq("myprofile")
      expect(resolved.token_duration).to eq(300)
      expect(resolved.pool_size).to eq(10)
      expect(resolved.application_name).to eq("rails")
    end

    it "returns a frozen ResolvedConfig" do
      config = described_class.new(host: "mycluster.dsql.us-east-1.on.aws")
      resolved = config.resolve

      expect(resolved).to be_frozen
      expect(resolved).to be_a(AuroraDsql::Pg::ResolvedConfig)
    end

    it "does not mutate original config" do
      config = described_class.new(host: "ijsamhssbh36dopuigphknejb4", region: "us-east-1")
      config.resolve

      expect(config.host).to eq("ijsamhssbh36dopuigphknejb4")
    end
  end

  describe ".parse" do
    it "parses basic connection string" do
      config = described_class.parse("postgres://admin@mycluster.dsql.us-east-1.on.aws/postgres")

      expect(config.host).to eq("mycluster.dsql.us-east-1.on.aws")
      expect(config.user).to eq("admin")
      expect(config.database).to eq("postgres")
    end

    it "parses connection string with port" do
      config = described_class.parse("postgres://admin@mycluster.dsql.us-east-1.on.aws:5433/mydb")

      expect(config.port).to eq(5433)
      expect(config.database).to eq("mydb")
    end

    it "parses DSQL-specific query params" do
      config = described_class.parse(
        "postgres://admin@mycluster.dsql.us-east-1.on.aws/postgres?region=eu-west-1&profile=dev&tokenDurationSecs=300"
      )

      expect(config.region).to eq("eu-west-1")
      expect(config.profile).to eq("dev")
      expect(config.token_duration).to eq(300)
    end

    it "handles postgresql:// scheme" do
      config = described_class.parse("postgresql://admin@mycluster.dsql.us-east-1.on.aws/postgres")

      expect(config.host).to eq("mycluster.dsql.us-east-1.on.aws")
    end
  end
end

RSpec.describe AuroraDsql::Pg::ResolvedConfig do
  let(:resolved) do
    AuroraDsql::Pg::ResolvedConfig.new(
      host: "mycluster.dsql.us-east-1.on.aws",
      region: "us-east-1",
      user: "admin",
      database: "postgres",
      port: 5432,
      profile: nil,
      token_duration: 900,
      credentials_provider: nil,
      max_lifetime: 3300,
      pool_size: 5,
      application_name: nil
    )
  end

  describe "#to_pg_params" do
    before do
      # Mock PG.library_version
      stub_const("PG", Class.new do
        def self.library_version
          170000
        end
      end)
    end

    it "returns hash with required pg connection params" do
      params = resolved.to_pg_params(password: "token123")

      expect(params[:host]).to eq("mycluster.dsql.us-east-1.on.aws")
      expect(params[:port]).to eq(5432)
      expect(params[:user]).to eq("admin")
      expect(params[:dbname]).to eq("postgres")
      expect(params[:password]).to eq("token123")
    end

    it "sets sslmode to verify-full" do
      params = resolved.to_pg_params(password: "token123")

      expect(params[:sslmode]).to eq("verify-full")
    end

    it "sets sslnegotiation to direct for libpq >= 17" do
      params = resolved.to_pg_params(password: "token123")

      expect(params[:sslnegotiation]).to eq("direct")
    end

    it "omits sslnegotiation for libpq < 17" do
      stub_const("PG", Class.new do
        def self.library_version
          160000
        end
      end)

      params = resolved.to_pg_params(password: "token123")

      expect(params).not_to have_key(:sslnegotiation)
    end

    it "sets application_name using build_application_name" do
      params = resolved.to_pg_params(password: "token123")

      expect(params[:application_name]).to eq("aurora-dsql-ruby-pg/1.0.0")
    end

    it "includes ORM prefix in application_name" do
      resolved_with_orm = AuroraDsql::Pg::ResolvedConfig.new(
        host: "mycluster.dsql.us-east-1.on.aws",
        region: "us-east-1",
        user: "admin",
        database: "postgres",
        port: 5432,
        profile: nil,
        token_duration: 900,
        credentials_provider: nil,
        max_lifetime: 3300,
        pool_size: 5,
        application_name: "rails"
      )

      params = resolved_with_orm.to_pg_params(password: "token123")

      expect(params[:application_name]).to eq("rails:aurora-dsql-ruby-pg/1.0.0")
    end
  end
end
